backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
-                        timeout
-                    ))
-                    .bind(job_id)
-                    .execute(db)
-                    .await;
+/// takes stdout and stderr from Child, panics if either are not present
+///
+/// builds a stream joining both stdout and stderr each read line by line
+fn child_joined_output_stream(
+    child: &mut Child,
+) -> impl futures::stream::FusedStream<Item = io::Result<String>> {
+    let stderr = child
+        .stderr
+        .take()
+        .expect("child did not have a handle to stdout");
 
-                    if q.is_err() {
-                        tracing::error!("error setting canceled for id {}", job_id);
-                    }
-                }
-                last_update = chrono::Utc::now().timestamp_millis();
-            },
-            nl = rx.recv() => {
+    let stdout = child
+        .stdout
+        .take()
+        .expect("child did not have a handle to stdout");
 
-                if let Some(nl) = nl {
-                    if logs.chars().count() > MAX_LOG_SIZE as usize{
-                        tracing::info!("Too many logs lines: {}", job_id);
-                        logs.push_str("Too many logs lines. Limit is 200000 chars. Killing job.");
-                        done.store(true, Ordering::Relaxed);
-                    }
+    let stdout = BufReader::new(stdout).lines();
+    let stderr = BufReader::new(stderr).lines();
+    futures::stream::select(lines_to_stream(stderr), lines_to_stream(stdout))
+}
 
-                    logs.push('\n');
-                    logs.push_str(&nl);
+fn lines_to_stream<R: tokio::io::AsyncBufRead + Unpin>(
+    mut lines: tokio::io::Lines<R>,
+) -> impl futures::Stream<Item = io::Result<String>> {
+    futures::stream::poll_fn(move |cx| {
+        std::pin::Pin::new(&mut lines)
+            .poll_next_line(cx)
+            .map(|result| result.transpose())
+    })
+}
 
-                    *last_line = nl;
-                } else {
-                    let to_send = logs.chars().skip(start).collect::<String>();
-                    concat_logs(&to_send, &job_id, db).await;
-                    break;
-                }
-            },
-        }
+// as a detail, `BufReader::lines()` removes \n and \r\n from the strings it yields,
+// so this pushes \n to thd destination string in each call
+fn append_with_limit(dst: &mut String, src: &str, limit: &mut usize) {
+    if *limit > 0 {
+        dst.push('\n');
     }
+    *limit -= 1;
 
-    let status = handle
-        .await
-        .map_err(|e| Error::ExecutionErr(e.to_string()))??;
-    Ok(status)
+    let src_len = src.chars().count();
+    if src_len <= *limit {
+        dst.push_str(&src);
+        *limit -= src_len;
+    } else {
+        let until = src
+            .char_indices()
+            .skip(*limit)
+            .next()
+            .map(|(until, _)| until)
+            .unwrap_or(0);
+        dst.push_str(&src[0..until]);
```

Comment:
I might be missing something obvious about usize conversion but why is it not `src[0..*limit]` or similar ?